src/charm.py

```diff
@@ -41,6 +42,48 @@
 CREATED_BY_LABEL = "gateway-api-integrator.charm.juju.is/managed-by"
 
 
+@lru_cache(maxsize=1)
+def get_client(field_manager: str, namespace: str) -> Client:
+    """Initialize the lightkube client with the correct namespace and field_manager.
+
+    Args:
+        field_manager (str): field manager for server side apply when patching resources.
+        namespace (str): The k8s namespace in which resources are managed.
+
+    Raises:
+        LightKubeInitializationError: When initialization of the lightkube client fails
+
+    Returns:
+        Client: The initialized lightkube client
+    """
+    try:
+        # Set field_manager for server-side apply when patching resources
+        # Keep this consistent across client initializations
+        kubeconfig = KubeConfig.from_service_account()
+        client = Client(config=kubeconfig, field_manager=field_manager, namespace=namespace)
+    except ConfigError as exc:
+        logger.exception("Error initializing the lightkube client.")
+        raise LightKubeInitializationError("Error initializing the lightkube client.") from exc
+
+    return client
+
+
+class LightKubeInitializationError(Exception):
+    """Exception raised when initialization of the lightkube client failed.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the InvalidCharmConfigError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
```

Comment:
I do not really see the benefit of this attribute as you can usually get the same effect using `str(e)`. Does also apply to other exceptions defined in this PR.

Reply:
I was thinking about this as well but did it this was as most of our charms implement errors like this. I removed this in relevant classes.